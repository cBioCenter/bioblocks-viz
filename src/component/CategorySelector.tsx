import * as React from 'react';
import { Dropdown } from 'semantic-ui-react';

export interface ICategorySelectorProps {
  /** Categories to select from. */
  categories: string[];
  /** Callback for when a new category is selected. */
  onCategoryChange?: (event: React.SyntheticEvent<any>, data: object) => void;
}

/**
 * Class to represent a dropdown.
 *
 * @extends {React.Component<ICategorySelectorProps, any>}
 */
export default class CategorySelector extends React.Component<ICategorySelectorProps, any> {
  constructor(props: ICategorySelectorProps) {
    super(props);
  }

  public render() {
    return (
      this.props.categories && (
        <Dropdown
          fluid={true}
          onChange={this.props.onCategoryChange}
          options={[
            { key: 'all', text: 'all', value: undefined },
            ...this.props.categories.map(cat => {
              return { key: cat, text: cat, value: cat };
            }),
          ]}
          placeholder={'Select Category'}
          search={true}
        />
      )
    );
  }
}

export { CategorySelector };
