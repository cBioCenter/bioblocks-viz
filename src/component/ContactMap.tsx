import * as React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';

import { ButtonProps, Icon } from 'semantic-ui-react';
import { createContainerActions, createResiduePairActions } from '~bioblocks-viz~/action';
import {
  ComponentCard,
  ContactMapChart,
  generateChartDataEntry,
  IComponentMenuBarItem,
  IContactMapChartData,
  IContactMapChartPoint,
} from '~bioblocks-viz~/component';
import {
  BIOBLOCKS_CSS_STYLE,
  BioblocksChartEvent,
  BioblocksWidgetConfig,
  ButtonGroupWidgetConfig,
  CONFIGURATION_COMPONENT_TYPE,
  CouplingContainer,
  IContactMapData,
  ICouplingScore,
  RESIDUE_TYPE,
  SECONDARY_STRUCTURE,
  SECONDARY_STRUCTURE_KEYS,
  SECONDARY_STRUCTURE_SECTION,
} from '~bioblocks-viz~/data';
import { ColorMapper, EMPTY_FUNCTION, generateCouplingScoreHoverText } from '~bioblocks-viz~/helper';
import { ILockedResiduePair } from '~bioblocks-viz~/reducer';
import { getCandidates, getHovered, getLocked, selectCurrentItems } from '~bioblocks-viz~/selector';

export type CONTACT_MAP_CB_RESULT_TYPE = ICouplingScore;
export type ContactMapCallback = (coupling: CONTACT_MAP_CB_RESULT_TYPE) => void;

export interface IContactMapProps {
  candidateResidues: RESIDUE_TYPE[];
  configurations: BioblocksWidgetConfig[];
  data: IContactMapData;
  formattedPoints: IContactMapChartData[];
  height: number | string;
  highlightColor: string;
  hoveredResidues: RESIDUE_TYPE[];
  hoveredSecondaryStructures: SECONDARY_STRUCTURE_SECTION[];
  isDataLoading: boolean;
  lockedResiduePairs: ILockedResiduePair;
  observedColor: string;
  secondaryStructureColors?: ColorMapper<SECONDARY_STRUCTURE_KEYS>;
  selectedSecondaryStructures: SECONDARY_STRUCTURE;
  showConfigurations: boolean;
  style?: BIOBLOCKS_CSS_STYLE;
  width: number | string;
  addHoveredResidues(residues: RESIDUE_TYPE[]): void;
  addHoveredSecondaryStructure(section: SECONDARY_STRUCTURE_SECTION): void;
  addSelectedSecondaryStructure(section: SECONDARY_STRUCTURE_SECTION): void;
  removeAllLockedResiduePairs(): void;
  removeAllSelectedSecondaryStructures(): void;
  removeSecondaryStructure(section: SECONDARY_STRUCTURE_SECTION): void;
  removeHoveredResidues(): void;
  removeHoveredSecondaryStructure(section: SECONDARY_STRUCTURE_SECTION): void;
  onBoxSelection?(residues: RESIDUE_TYPE[]): void;
  toggleLockedResiduePair(residuePair: ILockedResiduePair): void;
}

export const initialContactMapState = {
  pointsToPlot: new Array<IContactMapChartData>(),
};

export type ContactMapState = Readonly<typeof initialContactMapState>;

export class ContactMapClass extends React.Component<IContactMapProps, ContactMapState> {
  public static defaultProps = {
    addHoveredResidues: EMPTY_FUNCTION,
    addHoveredSecondaryStructure: EMPTY_FUNCTION,
    addSelectedSecondaryStructure: EMPTY_FUNCTION,
    candidateResidues: [],
    configurations: new Array<BioblocksWidgetConfig>(),
    data: {
      couplingScores: new CouplingContainer(),
      pdbData: { experimental: undefined, predicted: undefined },
      secondaryStructures: new Array<SECONDARY_STRUCTURE>(),
    },
    enableSliders: true,
    formattedPoints: new Array<IContactMapChartData>(),
    height: '100%',
    highlightColor: '#ff8800',
    hoveredResidues: [],
    hoveredSecondaryStructures: [],
    isDataLoading: false,
    lockedResiduePairs: {},
    observedColor: '#0000ff',
    removeAllLockedResiduePairs: EMPTY_FUNCTION,
    removeAllSelectedSecondaryStructures: EMPTY_FUNCTION,
    removeHoveredResidues: EMPTY_FUNCTION,
    removeHoveredSecondaryStructure: EMPTY_FUNCTION,
    removeSecondaryStructure: EMPTY_FUNCTION,
    selectedSecondaryStructures: [],
    showConfigurations: true,
    toggleLockedResiduePair: EMPTY_FUNCTION,
    width: '100%',
  };

  public readonly state: ContactMapState = initialContactMapState;

  constructor(props: IContactMapProps) {
    super(props);
  }

  public componentDidMount() {
    this.setupPointsToPlot(this.props.data.couplingScores);
  }

  public componentDidUpdate(prevProps: IContactMapProps) {
    const { data, lockedResiduePairs } = this.props;
    if (data !== prevProps.data || lockedResiduePairs !== prevProps.lockedResiduePairs) {
      this.setupPointsToPlot(data.couplingScores);
    }
  }

  public onNodeSizeChange = (index: number, nodeSizeMod: number) => (
    event?: React.MouseEvent<HTMLButtonElement>,
    data?: ButtonProps,
  ) => {
    const { pointsToPlot } = this.state;

    this.setState({
      pointsToPlot: [
        ...pointsToPlot.slice(0, index),
        {
          ...pointsToPlot[index],
          // This ensures a number in the range [1, 10]. 0 is an invalid point size in plotly.
          nodeSize: Math.min(Math.max(Math.abs(pointsToPlot[index].nodeSize + nodeSizeMod), 1), 10),
        },
        ...pointsToPlot.slice(index + 1),
      ],
    });
  };

  public render() {
    const { pointsToPlot } = this.state;

    return this.renderContactMapChart(pointsToPlot);
  }

  /**
   * Given a chart data entry, sets the node size if one already is set for this point.
   *
   * @param chartDatum A single chart data entry.
   */
  protected applyNodeSize = (chartDatum: IContactMapChartData) => {
    const { pointsToPlot } = this.state;
    const index = pointsToPlot.findIndex(currentPoint => currentPoint.name === chartDatum.name);
    if (index >= 0) {
      return {
        ...chartDatum,
        nodeSize: pointsToPlot[index].nodeSize,
      };
    } else {
      return chartDatum;
    }
  };

  /**
   * Returns an {i, j} pair for a hovered residue. Handles case when a single residue is hovered.
   *
   * @param hoveredResidues Array of residues that are hovered.
   */
  protected generateHoveredResiduePairs = (hoveredResidues: RESIDUE_TYPE[]) => {
    if (hoveredResidues.length >= 1) {
      return [
        {
          i: hoveredResidues[0],
          j: hoveredResidues.length === 1 ? hoveredResidues[0] : hoveredResidues[1],
        },
      ];
    }

    return [];
  };

  /**
   * Returns contact map points for locked residue pairs.
   *
   * @param hoveredResidues Array of residues that are hovered.
   */
  protected generateLockedResiduePairs = (lockedResiduePairs: ILockedResiduePair) => {
    const lockedResiduePairKeys = Object.keys(lockedResiduePairs);

    return lockedResiduePairKeys.reduce((reduceResult: IContactMapChartPoint[], key) => {
      const keyPair = lockedResiduePairs[key];
      if (keyPair && keyPair.length === 2) {
        reduceResult.push({ i: keyPair[0], j: keyPair[1], dist: 0 });
      }

      return reduceResult;
    }, new Array<IContactMapChartPoint>());
  };

  protected generateSelectedResiduesChartData = (
    highlightColor: string,
    chartName: string,
    nodeSize: number,
    chartPoints: IContactMapChartPoint[],
  ) => {
    return generateChartDataEntry('none', highlightColor, chartName, '', nodeSize, chartPoints, {
      marker: {
        color: new Array<string>(chartPoints.length * 2).fill(highlightColor),
        line: {
          color: highlightColor,
          width: 3,
        },
        symbol: 'circle-open',
      },
    });
  };

  /**
   * Gets the color from the provided chart entry.
   *
   * The following are checked in order - the first one found is the color returned:
   * - Marker's line color.
   * - Marker's color.
   * - Marker's colorscale.
   * - Line color.
   */
  protected getColorFromEntry = (entry: IContactMapChartData) => {
    if (entry.marker) {
      if (entry.marker.line && entry.marker.line.color) {
        return entry.marker.line.color;
      }
      if (entry.marker.color) {
        return Array.isArray(entry.marker.color) ? entry.marker.color[0] : entry.marker.color;
      }
      if (entry.marker.colorscale) {
        return Array.isArray(entry.marker.colorscale) ? entry.marker.colorscale[0][1] : entry.marker.colorscale;
      }
    }
    if (entry.line && entry.line.color) {
      return entry.line.color;
    }
  };

  protected getDockConfigs = () => [
    {
      isVisibleCb: () =>
        this.props.selectedSecondaryStructures.length >= 1 || Object.values(this.props.lockedResiduePairs).length >= 1,
      onClick: () => {
        this.props.removeAllLockedResiduePairs();
        this.props.removeAllSelectedSecondaryStructures();
        this.props.removeHoveredResidues();
      },
      text: 'Clear Selections',
    },
  ];

  protected getMenuConfigs = (
    Filters: BioblocksWidgetConfig[],
    pointsToPlot: IContactMapChartData[],
  ): IComponentMenuBarItem[] => [
    {
      component: {
        configs: { Filters },
        name: 'POPUP',
      },
      description: 'Filter',
      iconName: 'filter',
    },
    {
      component: {
        configs: this.getNodeSizeSliderConfigs(pointsToPlot),
        name: 'POPUP',
      },
      description: 'Settings',
    },
  ];

  protected getNodeSizeSliderConfigs = (entries: IContactMapChartData[]) => ({
    'Node Sizes': entries.map(
      (entry, index): ButtonGroupWidgetConfig => {
        const color = this.getColorFromEntry(entry);

        // prettier-ignore
        const options = [(
          <Icon
            key={`node-size-slider-${index}-minus`}
            name={'minus'}
            onClick={this.onNodeSizeChange(index, -1)}
            size={'small'}
            style={{ color }}
          />), (
          <Icon
            key={`node-size-slider-${index}-2`}
            name={'circle'}
            size={'small'}
            style={{ color }}
          />), (
          <Icon
            key={`node-size-slider-${index}-plus`}
            name={'plus'}
            onClick={this.onNodeSizeChange(index, 1)}
            size={'small'}
            style={{ color }}
          />
        )];

        return {
          id: `node-size-slider-${index}`,
          name: entry.name,
          options,
          style: { padding: '.5em' },
          type: CONFIGURATION_COMPONENT_TYPE.BUTTON_GROUP,
        };
      },
    ),
  });

  protected handleAxisClick = (e: BioblocksChartEvent) => {
    const { addSelectedSecondaryStructure, data, removeSecondaryStructure, selectedSecondaryStructures } = this.props;

    for (const secondaryStructure of data.secondaryStructures) {
      for (const section of secondaryStructure) {
        if (section.contains(...e.selectedPoints)) {
          if (selectedSecondaryStructures.includes(section)) {
            removeSecondaryStructure(section);
          } else {
            addSelectedSecondaryStructure(section);
          }
        }
      }
    }
  };

  protected handleAxisEnter = (e: BioblocksChartEvent) => {
    const { addHoveredSecondaryStructure, hoveredSecondaryStructures, data } = this.props;

    for (const secondaryStructure of data.secondaryStructures) {
      for (const section of secondaryStructure) {
        if (
          section.contains(...e.selectedPoints) &&
          !hoveredSecondaryStructures.find(secStruct => section.isEqual(secStruct))
        ) {
          addHoveredSecondaryStructure(section);
        }
      }
    }
  };

  protected handleAxisLeave = (e: BioblocksChartEvent) => {
    const { data, hoveredSecondaryStructures, removeHoveredSecondaryStructure } = this.props;

    for (const secondaryStructure of data.secondaryStructures) {
      for (const section of secondaryStructure) {
        const searchResult = hoveredSecondaryStructures.find(secStruct => section.isEqual(secStruct));
        if (section.contains(...e.selectedPoints) && searchResult) {
          removeHoveredSecondaryStructure(searchResult);
        }
      }
    }
  };

  protected onMouseClick = (cb: (residues: ILockedResiduePair) => void) => (e: BioblocksChartEvent) => {
    if (e.isAxis()) {
      this.handleAxisClick(e);
    } else {
      cb({ [e.selectedPoints.sort().toString()]: e.selectedPoints });
    }
  };

  protected onMouseEnter = (cb: (residue: RESIDUE_TYPE[]) => void) => (e: BioblocksChartEvent) => {
    if (e.isAxis()) {
      this.handleAxisEnter(e);
    } else {
      cb(e.selectedPoints);
    }
  };

  protected onMouseLeave = (cb?: (residue: RESIDUE_TYPE[]) => void) => (e: BioblocksChartEvent) => {
    if (e.isAxis()) {
      this.handleAxisLeave(e);
    } else if (cb) {
      cb(e.selectedPoints);
    }
  };

  protected onMouseSelect = (cb?: (residues: RESIDUE_TYPE[]) => void) => (e: BioblocksChartEvent) => {
    if (cb) {
      // For the contact map, all the x/y values are mirrored and correspond directly with i/j values.
      // Thus, all the residue numbers can be obtained by getting either all x or values from ths selected points.
      cb(e.selectedPoints.map(point => point));
    }
  };

  protected renderContactMapChart(pointsToPlot: IContactMapChartData[]) {
    const {
      addHoveredResidues,
      candidateResidues,
      configurations,
      data,
      height,
      isDataLoading,
      onBoxSelection,
      removeHoveredResidues,
      secondaryStructureColors,
      selectedSecondaryStructures,
      showConfigurations,
      toggleLockedResiduePair,
      width,
    } = this.props;

    const range = data.couplingScores.residueIndexRange.max + data.couplingScores.residueIndexRange.min;

    return (
      <ComponentCard
        componentName={'Contact Map'}
        dockItems={this.getDockConfigs()}
        isDataReady={data.couplingScores.allContacts.length >= 1}
        menuItems={this.getMenuConfigs(configurations, pointsToPlot)}
      >
        <ContactMapChart
          candidateResidues={candidateResidues}
          contactData={pointsToPlot}
          height={height}
          isDataLoading={isDataLoading}
          onClickCallback={this.onMouseClick(toggleLockedResiduePair)}
          onHoverCallback={this.onMouseEnter(addHoveredResidues)}
          onSelectedCallback={this.onMouseSelect(onBoxSelection)}
          onUnHoverCallback={this.onMouseLeave(removeHoveredResidues)}
          range={range}
          secondaryStructures={data.secondaryStructures ? data.secondaryStructures : []}
          secondaryStructureColors={secondaryStructureColors}
          selectedSecondaryStructures={[selectedSecondaryStructures]}
          showConfigurations={showConfigurations}
          width={width}
        />
      </ComponentCard>
    );
  }

  protected setupPointsToPlot(couplingContainer: CouplingContainer) {
    const { data, lockedResiduePairs, hoveredResidues, formattedPoints, observedColor, highlightColor } = this.props;

    const chartNames = {
      selected: 'Selected Residue Pair',
      structure: `${data.pdbData ? (data.pdbData.experimental ? 'X-ray' : 'Inferred') : 'Unknown'} Structure Contact`,
    };

    const observedContactPoints = couplingContainer.getObservedContacts();
    const result = new Array<IContactMapChartData>(
      generateChartDataEntry(
        'text',
        { start: observedColor, end: 'rgb(100,177,200)' },
        `Structure Contact (${data.pdbData ? 'PDB' : 'Coupling Scores'})`,
        chartNames.structure,
        4,
        observedContactPoints,
        {
          text: observedContactPoints.map(generateCouplingScoreHoverText),
        },
      ),
      ...formattedPoints,
    );

    const chartPoints = new Array<IContactMapChartPoint>();

    chartPoints.push(...this.generateHoveredResiduePairs(hoveredResidues));
    chartPoints.push(...this.generateLockedResiduePairs(lockedResiduePairs));

    result.push(this.generateSelectedResiduesChartData(highlightColor, chartNames.selected, 4, chartPoints));

    this.setState({
      ...this.state,
      pointsToPlot: result.map(this.applyNodeSize),
    });
  }
}

const mapStateToProps = (state: { [key: string]: any }) => ({
  candidateResidues: getCandidates(state).toArray(),
  hoveredResidues: getHovered(state).toArray(),
  hoveredSecondaryStructures: selectCurrentItems<SECONDARY_STRUCTURE_SECTION>(
    state,
    'secondaryStructure/hovered',
  ).toArray(),
  lockedResiduePairs: getLocked(state).toJS(),
  selectedSecondaryStructures: selectCurrentItems<SECONDARY_STRUCTURE_SECTION>(
    state,
    'secondaryStructure/selected',
  ).toArray(),
});

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      addHoveredResidues: createResiduePairActions().hovered.set,
      addHoveredSecondaryStructure: createContainerActions('secondaryStructure/hovered').add,
      addSelectedSecondaryStructure: createContainerActions('secondaryStructure/selected').add,
      removeAllLockedResiduePairs: createResiduePairActions().locked.clear,
      removeAllSelectedSecondaryStructures: createContainerActions('secondaryStructure/selected').clear,
      removeHoveredResidues: createResiduePairActions().hovered.clear,
      removeHoveredSecondaryStructure: createContainerActions('secondaryStructure/hovered').remove,
      removeSecondaryStructure: createContainerActions('secondaryStructure/selected').remove,
      toggleLockedResiduePair: createResiduePairActions().locked.toggle,
    },
    dispatch,
  );

export const ContactMap = connect(
  mapStateToProps,
  mapDispatchToProps,
)(ContactMapClass);
