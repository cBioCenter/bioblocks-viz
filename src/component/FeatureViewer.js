"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = require("react");
var immutable_1 = require("immutable");
var component_1 = require("~bioblocks-viz~/component");
var data_1 = require("~bioblocks-viz~/data");
var FeatureViewer = /** @class */ (function (_super) {
    tslib_1.__extends(FeatureViewer, _super);
    function FeatureViewer(props) {
        var _this = _super.call(this, props) || this;
        _this.onFeatureHover = function (event) {
            var e_1, _a;
            var _b = _this.props, data = _b.data, getTextForHover = _b.getTextForHover;
            var hoveredFeatureIndex = -1;
            // TODO Handle vertical viewer, better selection logic.
            var xCoords = [event.selectedPoints[0], event.selectedPoints[2]];
            for (var i = 0; i < data.length; ++i) {
                try {
                    for (var xCoords_1 = (e_1 = void 0, tslib_1.__values(xCoords)), xCoords_1_1 = xCoords_1.next(); !xCoords_1_1.done; xCoords_1_1 = xCoords_1.next()) {
                        var xCoord = xCoords_1_1.value;
                        if (data[i].contains(xCoord)) {
                            hoveredFeatureIndex = i;
                        }
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (xCoords_1_1 && !xCoords_1_1.done && (_a = xCoords_1.return)) _a.call(xCoords_1);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
            }
            _this.setState({
                hoverAnnotationText: getTextForHover && hoveredFeatureIndex >= 0
                    ? getTextForHover(data[hoveredFeatureIndex].label, hoveredFeatureIndex)
                    : '',
                hoveredFeatureIndex: hoveredFeatureIndex,
            });
        };
        _this.onFeatureClick = function (event) {
            var _a = _this.props, data = _a.data, onClickCallback = _a.onClickCallback;
            var selectedFeatureIndices = _this.deriveFeatureIndices(data, event.selectedPoints);
            if (onClickCallback) {
                onClickCallback(_this.deriveSelectedFeatures(data, selectedFeatureIndices.toArray()));
            }
            _this.setState({
                selectedFeatureIndices: selectedFeatureIndices,
                selectedRange: new data_1.Bioblocks1DSection('selection', -1, -1),
            });
        };
        _this.onFeatureSelect = function (event) {
            var _a = _this.props, data = _a.data, onSelectCallback = _a.onSelectCallback;
            var selectedFeatureIndices = _this.deriveFeatureIndices(data, event.selectedPoints);
            var plotlyEvent = event.plotlyEvent;
            var selectedRange = _this.state.selectedRange;
            if (plotlyEvent.range) {
                selectedRange = new data_1.Bioblocks1DSection('selection', Math.floor(plotlyEvent.range.x[0]), Math.ceil(plotlyEvent.range.x[1]));
                var selection = {
                    end: selectedRange.end,
                    featuresSelected: _this.deriveSelectedFeatures(data, selectedFeatureIndices.toArray()),
                    length: selectedRange.length,
                    start: selectedRange.start,
                };
                if (onSelectCallback) {
                    onSelectCallback(selection);
                }
            }
            _this.setState({
                selectedFeatureIndices: selectedFeatureIndices,
                selectedRange: selectedRange,
            });
        };
        /**
         * Derive the indices of the Features from the points the user selected.
         */
        _this.deriveFeatureIndices = function (data, userSelectedPoints) {
            var featureIndices = immutable_1.Set();
            var _loop_1 = function (i) {
                var xCoord = userSelectedPoints[i];
                data
                    .reduce(function (result, datum, index) { return (datum.contains(xCoord) ? tslib_1.__spread(result, [index]) : result); }, new Array())
                    .forEach(function (indexToAdd) {
                    featureIndices = featureIndices.add(indexToAdd);
                });
            };
            // Points come to us as [x0, y0, x1, y1, ..., xn, yn], so we skip every other point.
            for (var i = 0; i < userSelectedPoints.length; i += 2) {
                _loop_1(i);
            }
            return featureIndices;
        };
        /**
         * Shorthand to get the raw section data for a set of Features given some indices.
         */
        _this.deriveSelectedFeatures = function (data, selectedFeatureIndices) {
            return selectedFeatureIndices.map(function (index) { return new data_1.Bioblocks1DSection(data[index].label, data[index].start, data[index].end); });
        };
        _this.state = {
            hoverAnnotationText: '',
            hoveredFeatureIndex: -1,
            plotlyConfig: {
                showAxisDragHandles: false,
                showAxisRangeEntryBoxes: false,
            },
            plotlyData: [],
            plotlyLayout: {},
            selectedFeatureIndices: immutable_1.Set(),
            selectedRange: new data_1.Bioblocks1DSection('selection', -1, -1),
        };
        return _this;
    }
    FeatureViewer.getDerivedStateFromProps = function (nextProps, nextState) {
        var backgroundBar = nextProps.backgroundBar, data = nextProps.data, height = nextProps.height, maxLength = nextProps.maxLength, showGrouped = nextProps.showGrouped, title = nextProps.title, width = nextProps.width;
        var hoverAnnotationText = nextState.hoverAnnotationText, hoveredFeatureIndex = nextState.hoveredFeatureIndex, selectedRange = nextState.selectedRange;
        var maxGroups = -1;
        var groupData = data.map(function (datum, index) {
            var yIndex = showGrouped
                ? index
                : data.findIndex(function (candidateDatum) { return datum.label.localeCompare(candidateDatum.label) === 0; });
            maxGroups = Math.max(maxGroups, yIndex);
            return FeatureViewer.getPlotlyDataObject(datum, showGrouped, yIndex);
        });
        var backgroundBarData = FeatureViewer.getPlotlyBackgroundBarObject(backgroundBar
            ? backgroundBar
            : new data_1.TintedBioblocks1DSection('', 0, data.reduce(function (prev, cur) { return Math.max(prev, cur.end); }, -1), '#b9bcb6'), showGrouped, maxGroups / 2);
        var selectionData = {
            fill: 'toself',
            hoverinfo: 'none',
            line: {
                width: 0,
            },
            marker: {
                color: '#f9f69f',
                opacity: 0.5,
            },
            mode: 'lines+markers',
            showlegend: false,
            x: selectedRange.start !== -1 && selectedRange.end !== -1
                ? [selectedRange.start, selectedRange.end, selectedRange.end, selectedRange.start]
                : [],
            y: [0, 0, maxGroups + 1, maxGroups + 1],
        };
        var plotlyData = tslib_1.__spread([backgroundBarData, selectionData], groupData);
        return {
            plotlyData: plotlyData,
            plotlyLayout: {
                annotations: FeatureViewer.getAnnotationPlotlyData(hoveredFeatureIndex, hoverAnnotationText, plotlyData[hoveredFeatureIndex]),
                dragmode: 'select',
                height: height,
                hovermode: 'closest',
                margin: {
                    b: 30,
                    t: 60,
                },
                showlegend: false,
                title: title,
                width: width,
                xaxis: data.length > 0
                    ? {
                        autorange: false,
                        fixedrange: true,
                        range: [0, maxLength ? maxLength : data.reduce(function (prev, cur) { return Math.max(prev, cur.end); }, -1) + 200],
                        showgrid: false,
                        tick0: 0,
                        tickmode: 'auto',
                        ticks: 'outside',
                    }
                    : { visible: false },
                yaxis: {
                    autorange: false,
                    fixedrange: true,
                    range: [-0.25, showGrouped ? 2 : data.length],
                    visible: false,
                },
            },
        };
    };
    FeatureViewer.getBoxForBioblocksSection = function (datum) {
        return [
            datum.end - (datum.end - datum.start) / 2,
            null,
            datum.start,
            datum.start,
            datum.end,
            datum.end,
            datum.start,
        ];
    };
    FeatureViewer.prototype.render = function () {
        var _a = this.props, width = _a.width, height = _a.height;
        var _b = this.state, plotlyConfig = _b.plotlyConfig, plotlyData = _b.plotlyData, plotlyLayout = _b.plotlyLayout;
        return (React.createElement("div", { style: { height: height, width: width } },
            React.createElement(component_1.PlotlyChart, { config: plotlyConfig, data: plotlyData, layout: plotlyLayout, onClickCallback: this.onFeatureClick, onHoverCallback: this.onFeatureHover, onSelectedCallback: this.onFeatureSelect, showLoader: false })));
    };
    FeatureViewer.defaultProps = {
        data: [],
        height: 200,
        showGrouped: false,
        title: '',
        width: 600,
    };
    FeatureViewer.getAnnotationPlotlyData = function (hoveredFeatureIndex, text, hoveredDatum) {
        return hoveredFeatureIndex >= 0 && hoveredDatum.x && hoveredDatum.y
            ? [
                {
                    align: 'left',
                    arrowhead: 0,
                    arrowsize: 1,
                    arrowwidth: 1,
                    ax: 0,
                    ay: -25,
                    bgcolor: '#ffffff',
                    bordercolor: '#000000',
                    borderpad: 5,
                    showarrow: true,
                    text: text,
                    x: hoveredDatum.x[0],
                    xref: 'x',
                    y: hoveredDatum.y[hoveredDatum.y.length - 3],
                    yref: 'y',
                },
            ]
            : [];
    };
    FeatureViewer.getPlotlyBackgroundBarObject = function (datum, showGrouped, yIndex) { return (tslib_1.__assign(tslib_1.__assign({}, FeatureViewer.getPlotlyDataObject(datum, showGrouped, yIndex)), { y: showGrouped
            ? [0.25, null, 0.25, 0.75, 0.75, 0.25, 0.25]
            : [yIndex + 0.5, null, yIndex + 1, yIndex, yIndex, yIndex + 1, yIndex + 1] })); };
    FeatureViewer.getPlotlyDataObject = function (datum, showGrouped, yIndex) { return ({
        fill: 'toself',
        fillcolor: datum.color.toString(),
        hoverinfo: 'none',
        hoveron: 'fills',
        line: {
            width: 0,
        },
        mode: 'text+lines',
        name: "" + datum.label,
        text: [datum.label],
        textfont: { color: ['#FFFFFF'] },
        type: 'scatter',
        // Creates a 'box' so we can fill it and hover over it and add a point to the middle for the label.
        x: FeatureViewer.getBoxForBioblocksSection(datum),
        y: showGrouped
            ? [0.5, null, 0, 1, 1, 0, 0]
            : [yIndex + 0.5, null, yIndex + 1, yIndex, yIndex, yIndex + 1, yIndex + 1],
    }); };
    return FeatureViewer;
}(React.Component));
exports.FeatureViewer = FeatureViewer;
//# sourceMappingURL=FeatureViewer.js.map