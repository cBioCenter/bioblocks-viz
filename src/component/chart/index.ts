export * from './AuxiliaryAxis';
export * from './ContactMapChart';
export * from './PlotlyChart';
export * from './SecondaryStructureAxis';
