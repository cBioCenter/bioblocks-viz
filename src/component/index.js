"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
tslib_1.__exportStar(require("./chart"), exports);
tslib_1.__exportStar(require("./widget"), exports);
tslib_1.__exportStar(require("./ContactMapComponent"), exports);
tslib_1.__exportStar(require("./FeatureViewer"), exports);
tslib_1.__exportStar(require("./NGLComponent"), exports);
tslib_1.__exportStar(require("./TensorTComponent"), exports);
tslib_1.__exportStar(require("./UMAPVisualization"), exports);
//# sourceMappingURL=index.js.map