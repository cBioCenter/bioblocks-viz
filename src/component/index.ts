export * from './chart';
export * from './widget';

export * from './ContactMap';
export * from './FeatureViewer';
export * from './InfoPanel';
export * from './MorpheusComponent';
export * from './NGLComponent';
export * from './PredictedContactMap';
export * from './TComponent';
export * from './TensorTComponent';
export * from './TFrameComponent';
export * from './VizSelectorPanel';
