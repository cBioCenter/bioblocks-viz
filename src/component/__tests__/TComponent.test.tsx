import { shallow } from 'enzyme';
import * as React from 'react';

import { TComponent } from '~chell-viz~/component';

describe('TComponent', () => {
  it('Should match existing snapshot when given simple data.', () => {
    const data = [[1, 2], [3, 4], [5, 6]];
    expect(shallow(<TComponent data={data} />)).toMatchSnapshot();
  });
});
