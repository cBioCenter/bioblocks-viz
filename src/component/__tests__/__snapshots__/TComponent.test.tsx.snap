// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TComponent Should match existing snapshot when given simple data. 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
